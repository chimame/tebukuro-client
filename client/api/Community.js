import ApiRequest from './ApiRequest'

export const createCommunityHandler = (name, description) => {
  const communityParams = {
    name: name,
    description: description
  }
  return ApiRequest.post('/communities', communityParams)
}
