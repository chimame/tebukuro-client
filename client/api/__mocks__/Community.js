const path = require('path')
const Community = jest.genMockFromModule('../Community')

let mockResult = Object.create(null)
export const __setMockResult = (normal = true) => {
  mockResult = normal
}

export const showCommunity = (id) => mockResult ? Promise.resolve({message: "normal mock"}) : Promise.reject({message: "error mock"})
export const createCommunity = require('../Community').createCommunity
export const getCommunities = require('../Community').getCommunities
