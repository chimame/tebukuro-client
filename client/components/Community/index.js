import React, { Component } from 'react'

export default class Community extends Component {
  render() {
    return (
      <div>
        <p>name: { this.props.community.name }</p>
        <p>description: { this.props.community.description }</p>
      </div>
    )
  }
}

Community.propTypes = {
  community: React.PropTypes.object.isRequired
}
