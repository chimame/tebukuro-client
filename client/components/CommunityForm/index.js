import React, { Component } from 'react'

const onErrorComponent = (community) => {
  if (community.isError()) {
    let errors = []

    community.errors.forEach((error) => {
      errors.push(<p key={ error }>{error}</p>)
    })

    return <div className="error">{ errors }</div>
  }
}

export default class CommunityForm extends Component {
  onSubmitHandler(e) {
    e.preventDefault()
    const name = this.refs.inputName.value
    const description = this.refs.inputDescription.value
    this.props.onSubmit(name, description)
  }

  render() {
    return (
      <div>
        <h3>create community form</h3>
        { onErrorComponent(this.props.community) }
        <form onSubmit={ this.onSubmitHandler.bind(this) }>
          <div>
            <label>
              name
              <input type="text" ref="inputName" />
            </label>
          </div>
          <div>
            <label>
              description
              <textarea ref="inputDescription" />
            </label>
          </div>
          <div>
            <input type="submit" />
          </div>
        </form>
      </div>
    )
  }
}

CommunityForm.propTypes = {
  onSubmit: React.PropTypes.func.isRequired,
  community: React.PropTypes.object.isRequired
}
