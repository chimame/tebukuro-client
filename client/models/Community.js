import { Record } from 'immutable'

export const CommunityRecord = Record({
  id:           0,
  name:         '',
  description:  '',
  errors:       []
})

export default class Community extends CommunityRecord {
  isError() {
    return this.errors.length !== 0
  }
}
