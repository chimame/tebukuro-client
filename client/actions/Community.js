import { createAction } from 'redux-actions'
import { createCommunityHandler } from '../api/Community'

export const CREATE_COMMUNITY = 'CREATE_COMMUNITY'

export const createCommunity = createAction(CREATE_COMMUNITY, createCommunityHandler)
