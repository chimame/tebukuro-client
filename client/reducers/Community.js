import { handleActions }  from 'redux-actions'
import Community          from '../models/Community'

export const communityInitialState = new Community()

export const communityReducerMap = {
  CREATE_COMMUNITY: {
    next: (state, action) => {
      return new Community({
        id:           action.payload.id,
        name:         action.payload.name,
        description:  action.payload.description
      })
    },
    throw: (state, action) => {
      return new Community({
        errors: action.payload.messages
      })
    }
  }
}

export default handleActions(communityReducerMap, communityInitialState)
