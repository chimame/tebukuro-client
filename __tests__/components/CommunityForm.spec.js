import React from 'react'
import { mount } from 'enzyme';
import CommunityForm from '../../client/components/CommunityForm'
import Community from '../../client/models/Community'
import renderer from 'react-test-renderer'

describe('CommunityForm', () => {
  it('render form', () => {
    const onSubmitMock = jest.fn()

    const component = renderer.create(
      <CommunityForm onSubmit={onSubmitMock} community={new Community()}/>
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('props onSubmit', () => {
    let renderedForm
    let onSubmitMock

    beforeEach(() => {
      onSubmitMock = jest.fn()

      renderedForm = mount(
        <CommunityForm onSubmit={onSubmitMock} community={new Community()} />
      )
    })

    describe('when submit', () => {
      beforeEach(() => {
        renderedForm.find('[type="submit"]').get('0').click()
      })

      it('should execute onSubmit function', () => {
        expect(onSubmitMock).toBeCalled()
      })
    })
  })

  describe('when error', () => {
    let renderedForm

    beforeEach(() => {
      renderedForm = mount(
        <CommunityForm onSubmit={() => {}} community={new Community({errors: ['error']})} />
      )
    })

    it('should exists error claass dom', () => {
      expect(renderedForm.find('.error p').length).toEqual(1)
    })
  })
})
