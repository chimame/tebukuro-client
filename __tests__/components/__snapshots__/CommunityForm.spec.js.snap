exports[`CommunityForm render form 1`] = `
<div>
  <h3>
    create community form
  </h3>
  <form
    onSubmit={[Function]}>
    <div>
      <label>
        name
        <input
          type="text" />
      </label>
    </div>
    <div>
      <label>
        description
        <textarea />
      </label>
    </div>
    <div>
      <input
        type="submit" />
    </div>
  </form>
</div>
`;
